File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgProgramListRecyclerViewAdapter.java
package org.tvheadend.tvhclient.ui.features.epg;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.EpgProgram;
import org.tvheadend.tvhclient.domain.entity.Recording;
import org.tvheadend.tvhclient.databinding.EpgProgramItemAdapterBinding;
import org.tvheadend.tvhclient.ui.features.programs.ProgramDetailsFragment;
import org.tvheadend.tvhclient.ui.common.callbacks.RecyclerViewClickCallback;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

class EpgProgramListRecyclerViewAdapter extends RecyclerView.Adapter<EpgProgramListRecyclerViewAdapter.EpgProgramListViewHolder> implements RecyclerViewClickCallback {

    private final float pixelsPerMinute;
    private final long fragmentStartTime;
    private final long fragmentStopTime;
    private final List<EpgProgram> programList = new ArrayList<>();
    private List<Recording> recordingList = new ArrayList<>();

    EpgProgramListRecyclerViewAdapter(float pixelsPerMinute, long fragmentStartTime, long fragmentStopTime) {
        this.pixelsPerMinute = pixelsPerMinute;
        this.fragmentStartTime = fragmentStartTime;
        this.fragmentStopTime = fragmentStopTime;
    }

    @NonNull
    @Override
    public EpgProgramListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        EpgProgramItemAdapterBinding itemBinding = EpgProgramItemAdapterBinding.inflate(layoutInflater, parent, false);

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(parent.getContext());
        boolean showProgramSubtitle = sharedPreferences.getBoolean("program_subtitle_enabled", parent.getContext().getResources().getBoolean(R.bool.pref_default_program_subtitle_enabled));
        boolean showGenreColors = sharedPreferences.getBoolean("genre_colors_for_program_guide_enabled", parent.getContext().getResources().getBoolean(R.bool.pref_default_genre_colors_for_program_guide_enabled));

        return new EpgProgramListViewHolder(itemBinding, showProgramSubtitle, showGenreColors, this);
    }

    @Override
    public void onBindViewHolder(@NonNull EpgProgramListViewHolder holder, int position) {
        if (programList.size() > position) {
            EpgProgram program = programList.get(position);

            long startTime = (program.getStart() < fragmentStartTime) ? fragmentStartTime : program.getStart();
            long stopTime = (program.getStop() > fragmentStopTime) ? fragmentStopTime : program.getStop();
            int layoutWidth = (int) (((stopTime - startTime) / 1000 / 60) * pixelsPerMinute);

            holder.bind(program, position, layoutWidth);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull EpgProgramListViewHolder holder, int position, @NonNull List<Object> payloads) {
        onBindViewHolder(holder, position);
    }

    void addItems(@NonNull List<EpgProgram> newItems) {
        updateRecordingState(newItems, recordingList);

        List<EpgProgram> oldItems = new ArrayList<>(programList);
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new EpgProgramListDiffCallback(oldItems, newItems));

        programList.clear();
        programList.addAll(newItems);
        diffResult.dispatchUpdatesTo(this);
    }

    void addRecordings(@NonNull List<Recording> list) {
        recordingList.clear();
        recordingList.addAll(list);
        updateRecordingState(programList, recordingList);
    }

    private void updateRecordingState(@NonNull List<EpgProgram> programs, @NonNull List<Recording> recordings) {
        for (int i = 0; i < programs.size(); i++) {
            EpgProgram program = programs.get(i);
            boolean recordingExists = false;

            for (Recording recording : recordings) {
                if (program.getEventId() > 0 && program.getEventId() == recording.getEventId()) {
                    Recording oldRecording = program.getRecording();
                    program.setRecording(recording);

                    // Do a full update only when a new recording was added or the recording
                    // state has changed which results in a different recording state icon
                    // Otherwise do not update the UI
                    if (oldRecording == null
                            || (!TextUtils.equals(oldRecording.getError(), recording.getError())
                            || !TextUtils.equals(oldRecording.getState(), recording.getState()))) {
                        notifyItemChanged(i);
                    }
                    recordingExists = true;
                    break;
                }
            }
            if (!recordingExists && program.getRecording() != null) {
                program.setRecording(null);
                notifyItemChanged(i);
            }
            programs.set(i, program);
        }
    }

    @Override
    public int getItemCount() {
        return programList.size();
    }

    @Override
    public int getItemViewType(final int position) {
        return R.layout.epg_program_item_adapter;
    }

    private EpgProgram getItem(int position) {
        if (programList.size() > position && position >= 0) {
            return programList.get(position);
        } else {
            return null;
        }
    }

    /**
     * Returns the activity from the view context so that
     * stuff like the fragment manager can be accessed
     *
     * @param view The view to retrieve the activity from
     * @return Activity or null if none was found
     */
    private AppCompatActivity getActivity(View view) {
        Context context = view.getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof AppCompatActivity) {
                return (AppCompatActivity) context;
            }
            context = ((ContextWrapper) context).getBaseContext();
        }
        return null;
    }

    @Override
    public void onClick(@NonNull View view, int position) {
        AppCompatActivity activity = getActivity(view);
        EpgProgram program = getItem(position);
        if (program == null || activity == null) {
            return;
        }

        Fragment fragment = ProgramDetailsFragment.newInstance(program.getEventId(), program.getChannelId());
        FragmentTransaction ft = activity.getSupportFragmentManager().beginTransaction();
        ft.replace(R.id.main, fragment);
        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        ft.addToBackStack(null);
        ft.commit();
    }

    @Override
    public boolean onLongClick(@NonNull View view, int position) {
        AppCompatActivity activity = getActivity(view);
        EpgProgram program = getItem(position);
        if (program == null || activity == null) {
            return false;
        }

        Fragment fragment = activity.getSupportFragmentManager().findFragmentById(R.id.main);
        if (fragment instanceof ProgramGuideFragment
                && fragment.isAdded()
                && fragment.isResumed()) {
            ((ProgramGuideFragment) fragment).showPopupMenu(view, program);
        }
        return true;
    }

    static class EpgProgramListViewHolder extends RecyclerView.ViewHolder {

        private final EpgProgramItemAdapterBinding binding;
        private final boolean showProgramSubtitle;
        private final boolean showGenreColors;
        private final RecyclerViewClickCallback clickCallback;

        EpgProgramListViewHolder(EpgProgramItemAdapterBinding binding, boolean showProgramSubtitle, boolean showGenreColors, RecyclerViewClickCallback clickCallback) {
            super(binding.getRoot());
            this.binding = binding;
            this.showProgramSubtitle = showProgramSubtitle;
            this.showGenreColors = showGenreColors;
            this.clickCallback = clickCallback;
        }

        void bind(EpgProgram program, int position, int layoutWidth) {
            binding.setProgram(program);
            binding.setPosition(position);
            binding.setLayoutWidth(layoutWidth);
            binding.setShowGenreColor(showGenreColors);
            binding.setShowProgramSubtitle(showProgramSubtitle);
            binding.setCallback(clickCallback);
            binding.executePendingBindings();
        }
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgProgramListRecyclerViewAdapter.kt
package org.tvheadend.tvhclient.ui.features.epg

import android.content.ContextWrapper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.EpgProgramItemAdapterBinding
import org.tvheadend.tvhclient.domain.entity.EpgProgram
import org.tvheadend.tvhclient.domain.entity.Recording
import org.tvheadend.tvhclient.ui.common.callbacks.RecyclerViewClickCallback
import org.tvheadend.tvhclient.ui.features.programs.ProgramDetailsFragment
import org.tvheadend.tvhclient.util.isEqualTo
import java.util.*

internal class EpgProgramListRecyclerViewAdapter(private val viewModel: EpgViewModel, private val fragmentId: Int) : RecyclerView.Adapter<EpgProgramListRecyclerViewAdapter.EpgProgramListViewHolder>(), RecyclerViewClickCallback {

    private val programList = ArrayList<EpgProgram>()
    private val recordingList = ArrayList<Recording>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EpgProgramListViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val itemBinding = EpgProgramItemAdapterBinding.inflate(layoutInflater, parent, false)
        val viewHolder = EpgProgramListViewHolder(itemBinding, viewModel, this)
        itemBinding.lifecycleOwner = viewHolder
        return viewHolder
    }

    override fun onBindViewHolder(holder: EpgProgramListViewHolder, position: Int) {
        if (programList.size > position) {
            val program = programList[position]

            val startTime = if (program.start < viewModel.getStartTime(fragmentId)) viewModel.getStartTime(fragmentId) else program.start
            val stopTime = if (program.stop > viewModel.getEndTime(fragmentId)) viewModel.getEndTime(fragmentId) else program.stop
            val layoutWidth = ((stopTime - startTime) / 1000 / 60 * viewModel.pixelsPerMinute).toInt()

            holder.bind(program, position, layoutWidth)
        }
    }

    override fun onBindViewHolder(holder: EpgProgramListViewHolder, position: Int, payloads: List<Any>) {
        onBindViewHolder(holder, position)
    }

    fun addItems(newItems: MutableList<EpgProgram>) {
        updateRecordingState(newItems, recordingList)

        val oldItems = ArrayList(programList)
        val diffResult = DiffUtil.calculateDiff(EpgProgramListDiffCallback(oldItems, newItems))

        programList.clear()
        programList.addAll(newItems)
        diffResult.dispatchUpdatesTo(this)
    }

    fun addRecordings(list: List<Recording>) {
        recordingList.clear()
        recordingList.addAll(list)
        updateRecordingState(programList, recordingList)
    }

    private fun updateRecordingState(programs: MutableList<EpgProgram>, recordings: List<Recording>) {
        for (i in programs.indices) {
            val program = programs[i]
            var recordingExists = false

            for (recording in recordings) {
                if (program.eventId > 0 && program.eventId == recording.eventId) {
                    val oldRecording = program.recording
                    program.recording = recording

                    // Do a full update only when a new recording was added or the recording
                    // state has changed which results in a different recording state icon
                    // Otherwise do not update the UI
                    if (oldRecording == null
                            || !oldRecording.error.isEqualTo(recording.error)
                            || !oldRecording.state.isEqualTo(recording.state)) {
                        notifyItemChanged(i)
                    }
                    recordingExists = true
                    break
                }
            }
            if (!recordingExists && program.recording != null) {
                program.recording = null
                notifyItemChanged(i)
            }
            programs[i] = program
        }
    }

    override fun getItemCount(): Int {
        return programList.size
    }

    override fun getItemViewType(position: Int): Int {
        return R.layout.epg_program_item_adapter
    }

    private fun getItem(position: Int): EpgProgram? {
        return if (programList.size > position && position >= 0) {
            programList[position]
        } else {
            null
        }
    }

    /**
     * Returns the activity from the view context so that
     * stuff like the fragment manager can be accessed
     *
     * @param view The view to retrieve the activity from
     * @return Activity or null if none was found
     */
    private fun getActivity(view: View): AppCompatActivity? {
        var context = view.context
        while (context is ContextWrapper) {
            if (context is AppCompatActivity) {
                return context
            }
            context = context.baseContext
        }
        return null
    }

    override fun onClick(view: View, position: Int) {
        val activity = getActivity(view)
        val program = getItem(position)
        if (program == null || activity == null) {
            return
        }

        val fragment = ProgramDetailsFragment.newInstance(program.eventId, program.channelId)
        val ft = activity.supportFragmentManager.beginTransaction()
        ft.replace(R.id.main, fragment)
        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
        ft.addToBackStack(null)
        ft.commit()
    }

    override fun onLongClick(view: View, position: Int): Boolean {
        val activity = getActivity(view)
        val program = getItem(position)
        if (program == null || activity == null) {
            return false
        }

        val fragment = activity.supportFragmentManager.findFragmentById(R.id.main)
        if (fragment is ProgramGuideFragment
                && fragment.isAdded
                && fragment.isResumed) {
            fragment.showPopupMenu(view, program)
        }
        return true
    }

    override fun onViewAttachedToWindow(holder: EpgProgramListViewHolder) {
        super.onViewAttachedToWindow(holder)
        holder.markAttach()
    }

    override fun onViewDetachedFromWindow(holder: EpgProgramListViewHolder) {
        super.onViewDetachedFromWindow(holder)
        holder.markDetach()
    }

    internal class EpgProgramListViewHolder(private val binding: EpgProgramItemAdapterBinding, private val viewModel: EpgViewModel, private val clickCallback: RecyclerViewClickCallback) : RecyclerView.ViewHolder(binding.root), LifecycleOwner {

        private val lifecycleRegistry = LifecycleRegistry(this)

        init {
            lifecycleRegistry.currentState = Lifecycle.State.INITIALIZED
        }

        fun markAttach() {
            lifecycleRegistry.currentState = Lifecycle.State.STARTED
        }

        fun markDetach() {
            lifecycleRegistry.currentState = Lifecycle.State.DESTROYED
        }

        override fun getLifecycle(): Lifecycle {
            return lifecycleRegistry
        }

        fun bind(program: EpgProgram, position: Int, layoutWidth: Int) {
            binding.program = program
            binding.position = position
            binding.layoutWidth = layoutWidth
            binding.viewModel = viewModel
            binding.callback = clickCallback
            binding.executePendingBindings()
        }
    }
}
